### PR TITLE
a rosout publisher of a node might not exist

### LINKED ISSUE #1195: Exception when instantiating ActionServer as part of ROS package tests
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 22.04
- Installation type:
  - Binaries
- Version or commit hash:
  - Latest binaries of Humble/Iron/Rolling
- DDS implementation:
  - Fast DDS or Cyclone DDS (doesn't make a difference)
- Client library (if applicable):
  - rclpy

#### Steps to reproduce issue
I have a test that is failing as part of this PR (https://github.com/RobotWebTools/rosbridge_suite/pull/886) when I try to instantiate an action server, noting that:

* This only happens on Iron and Rolling, but not on Humble
* This only happens when I run using `colcon test` (which in turn uses `ament_add_pytest_test` CMake capability), but not if I run it directly using Python

Can't seem to figure out why this would be the case, so figured I'd elevate it here? Thank you!

The code:
```python
from example_interfaces.action import Fibonacci

class MyClass:
    def __init__(self, node):
        self.action_server = ActionServer(
            node,  # Just an rclpy.Node
            Fibonacci,
            "/fibonacci_action_1",
            self.execute_callback,
        )
```

The error:
```
          self.action_server = ActionServer(
      /opt/ros/rolling/lib/python3.10/site-packages/rclpy/action/server.py:264: in __init__
          self._logger = self._node.get_logger().get_child('action_server')
      _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
      
      self = <rclpy.impl.rcutils_logger.RcutilsLogger object at 0x7f730bfe9270>
      name = 'action_server'
      
          def get_child(self, name):
              if not name:
                  raise ValueError('Child logger name must not be empty.')
          
              if self.name:
                  # Prepend the name of this logger
      >           _rclpy.rclpy_logging_rosout_add_sublogger(self.name, name)
      E           RuntimeError: Failed to call rcl_logging_rosout_add_sublogger
```

#### Expected behavior
Should not raise an exception.

#### Actual behavior
Raises an exception.

### PR DESCRIPTION
**Pull request by iuhilnehc-ynos**
- [ ] https://github.com/ros2/rcl/pull/1115
- [ ] https://github.com/ros2/rclpy/pull/1196
- [ ] https://github.com/ros2/rclcpp/pull/2357

to fix https://github.com/ros2/rclpy/issues/1195

### CODE REVIEW

**Comment by iuhilnehc-ynos** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19714/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14224/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20443/)

File: rcl/src/rcl/logging_rosout.c [hunk 1]
```diff
@@ -438,20 +438,20 @@ rcl_logging_rosout_add_sublogger(
   rosout_map_entry_t entry;
   rosout_sublogger_entry_t sublogger_entry;
 
+  RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
```
Comment (fujitatomoya) [inline rcl/src/rcl/logging_rosout.c:443]: +1, these checks are good to add.

File: rcl/src/rcl/logging_rosout.c [hunk 2]
```diff
@@ -438,20 +438,20 @@ rcl_logging_rosout_add_sublogger(
...
+  rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
+  if (RCUTILS_RET_NOT_FOUND == rcutils_ret) {
+    return RCL_RET_OK;
+  }
```
Comment (fujitatomoya) [inline rcl/src/rcl/logging_rosout.c]: I am not sure if this is expected behavior... IMO this still needs to fail? cz it failed to add sub-logger to the hash map? this situation could happen with certain condition such as disabling rosout or node might not exist. no matter it is, it failed to add the sub-logger, so user should expect the failure for that?
how about dealing with case in client libraries, we can check the error code, and if that is RCUTILS_RET_NOT_FOUND with warning?

Comment (iuhilnehc-ynos) [reply rcl/src/rcl/logging_rosout.c]: Thanks.
What I did here is similar to https://github.com/ros2/rcl/blob/11103d750af1dc537d229a7e99bb05180c890088/rcl/src/rcl/logging_rosout.c#L431-L433 that if there is uninitialized.

I think your suggestion makes sense. I'll update it.

**Comment by sea-bass** [comment]
Thanks for looking at this fix so quickly!

Just leaving a comment here that this should also be backported to Iron.

**Comment by iuhilnehc-ynos** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19732/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14238/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20459/)

File: rcl/include/rcl/logging_rosout.h [hunk 3]
```diff
@@ -213,7 +213,7 @@ rcl_logging_rosout_output_handler(
  * \param[in] sublogger_name a sublogger name
  * \return #RCL_RET_OK if the subordinate logger was created successfully, or
  * \return #RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
- * \return #RCL_RET_SUBLOGGER_ALREADY_EXIST if the subordinate logger already exists, or
```
Comment (fujitatomoya) [inline rcl/include/rcl/logging_rosout.h:216]: this is not used, so we need to remove this.

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by iuhilnehc-ynos** [comment]
@clalancette Could you please help review it?

**Review by SteveMacenski — APPROVED** [review]

File: rcl/include/rcl/logging_rosout.h [hunk 4]
```diff
@@ -213,7 +213,7 @@ rcl_logging_rosout_output_handler(
...
+ * \return #RCL_RET_NOT_FOUND if the parent logger not exists, or
```
Comment (clalancette) [inline rcl/include/rcl/logging_rosout.h]:
```suggestion
 * \return #RCL_RET_NOT_FOUND if the parent logger does not exist, or
```

File: rcl/include/rcl/logging_rosout.h [hunk 5]
```diff
@@ -242,6 +242,7 @@ rcl_logging_rosout_add_sublogger(
  * \param[in] sublogger_name a sublogger name
  * \return #RCL_RET_OK if the subordinate logger was finalized successfully, or
  * \return #RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ * \return #RCL_RET_NOT_FOUND if the sublogger not exists, or
```
Comment (clalancette) [inline rcl/include/rcl/logging_rosout.h]:
```suggestion
 * \return #RCL_RET_NOT_FOUND if the sublogger does not exist, or
```

File: rcl/src/rcl/logging_rosout.c [hunk 6]
```diff
@@ -438,20 +439,23 @@ rcl_logging_rosout_add_sublogger(
...
   rosout_sublogger_entry_t sublogger_entry;
 
+  RCL_CHECK_ARGUMENT_FOR_NULL(logger_name, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(sublogger_name, RCL_RET_INVALID_ARGUMENT);
+  rcutils_ret_t rcutils_ret = rcutils_hash_map_get(&__logger_map, &logger_name, &entry);
+  if (RCL_RET_OK != (status = rcl_convert_rcutils_ret_to_rcl_ret(rcutils_ret))) {
+    if (RCL_RET_NOT_FOUND == status) {
+      RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to get logger entry for '%s'.", logger_name);
+    }
+    return status;
```
Comment (clalancette) [inline rcl/src/rcl/logging_rosout.c]: I think we can make this easier to read with:

```suggestion
  if (RCUTILS_RET_OK != rcutils_ret) {
    if (RCUTILS_RET_NOT_FOUND == status) {
      RCL_SET_ERROR_MSG_WITH_FORMAT_STRING("Failed to get logger entry for '%s'.", logger_name);
    }
    return rcl_convert_rcutils_ret_to_rcl_ret(rcutils_ret);
```

**Review by clalancette — CHANGES_REQUESTED** [review]
I've left a couple of things I think we can improve here, but overall it looks good.

**Comment by iuhilnehc-ynos** [comment]
> I've left a couple of things I think we can improve here, but overall it looks good.

Thank you. Updated in https://github.com/ros2/rcl/pull/1115/commits/b376e58cce3f038d7571cc750a33c6b0e5b9674a.

**Comment by iuhilnehc-ynos** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/19782/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14284/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20506/)

**Review by clalancette — APPROVED** [review]

**Pull request merged**